README
# check_temp

A small Nagios/Icinga plugin that checks the CPU (or M/B) temperature with lm-sensors.
It's written in Bash and uses *nix "sensors" and some sed & awk.
Default is to check the CPU temperature but this can be changed to, for example, 
the motherboard temperature with a "--sensor" argument. 
The plugin complies with the guidelines, for example uses -w -c -v arguments 
etc. It also does some basic sanity checks and has a exit 3 catch-all.

The plugin was submitted to Nagios Exchange in 2011.

Known forks of check_temp:
There is a very good Perl fork of check_temp written by Chad Columbus. It's
available on Nagios Exchange at
http://exchange.nagios.org/directory/Plugins/Operating-Systems/Linux/check_temp-2Epl/details

## Required
1) Install lm-sensors:
  a) On Debian/Ubuntu... apt-get install lm-sensors
  b) On OpenSUSE etc... zypper in sensors

2) Run sensors-detect:
  sudo sensors-detect

Let it check for any sensors that you feel are necessary, or all. You can just
press <ENTER> to have it use the default option for each check. There is a
warning in the manpages (man sensors-detect) that there are (some rare)
hardware sensors that may lock up or even be permanently damaged, so be aware
of that.

**********
MAKE SURE YOU TYPE "YES" TO THE LAST OPTION: "Do you want to add these lines automatically to /etc/modules? (yes/NO)"
**********
 
3) Restart module-init-tools service (On Debian/Ubuntu)
  a) /etc/init.d/module-init-tools restart
    or
  b) service module-init-tools restart

# Sensor 101

The output normally has a set of one or more blocks. This totally depends
on your hardware, available and loaded modules. A single block looks like
the following:

```
k10temp-pci-00c3
Adapter: PCI adapter
temp1:        +45.2°C  (high = +70.0°C)
                       (crit = +105.0°C, hyst = +97.5°C)
```

The first line is the label and represents a device. As shown above the
device is named "k10temp-pci-00c3". The second line can be suppressed
with `sensors -A`. It represents the type of adapter of the device. 
Each device has at least one sensor. In the above case the sensor name is
temp1. 

There can be other devices like a laptop battery. It could have only
a single sensor and it shows the battery's current voltage. A more complex
example with a battery a GPU and a CPU:

```
BAT1-acpi-0
in0:          15.54 V  

radeon-pci-0008
temp1:        +45.5°C  (crit = +120.0°C, hyst = +90.0°C)

k10temp-pci-00c3
temp1:        +44.1°C  (high = +70.0°C)
                       (crit = +105.0°C, hyst = +97.5°C)
```

This sample shows a quite common issue of `--sensor` without
`-n`: temp1 is not unique and the first occurence will become the
input for the monitoring system.

# Tests

Under "tests" is a script `run_tests.sh` that executes all test scripts in
random order. Each test script comes with a mocked output of sensors. It
plays tricks on `check_temp.sh`. It is not necessary to instrument any
binary nor `sudo` or symbolic links.

To create your test case:

1. Think of a good name for the use case the tests will assert, e.g. "new_filter"
1. Redirect the output to a data file: `sensors >tests/data/sample_new_filter`
1. copy one of the existing tests that suits your needs best, e.g. `test_new_filter.sh`
1. edit the skeleton to invoke check_temp.sh with the necessary parameters
1. file a pull request to outline the problem with the newly added test code

The common case is check_temp.sh without the -n switch which invokes sensors
command without any parameters. Use TESTDATA and a single file as mocked
sensors output.

If you intend to test with -n switch instead use TESTDATA_PREFIX. Each device
specified with --sensor is mapped to the mock file TESTDATA_PREFIX-<sensor>.
The test script `test_new_filter.sh` illustrates this.


HISTORY
Version 0.2: 
   Line 103, fixed the missing "-n" option (Thanks to Chad who pointed this out
   this out for me). Also added a "shopt -s extglob". (Thx to Chad)

Version 0.5: 
   Line 168, fixed a typo (EXIT_UNKNOWN to STATE_UNKNOWN)

Version 0.7:
   Line 193, modified sensor parsing to cut after the first '+' since all
     positive temperatures are preceded by a '+'
   Line 211, reduced "verbosity" needed to see verbose info (was 2: -v -v)
   Line 229-232, now checks to see if no sensor data was found and
     exits with STATE_UNKNOWN

Version 0.8:
   Line 196 and 230, added 'head -n1' to only fetch the first result from
     sensors output. On some machine you get two Core0 and two Core1 temps.
   Moved version history to it's own file, HISTORY

Version 0.94:
    Lots of changes and bugfixes. A huge thank you to Ryan Loudfoot who
    has fixed bugs and worked out a new way to detect the sensors-program.
    In 0.94 I removed /etc from the whereis path as it didn't work for me
    on my setup. It found the config-file for lm-sensors and tried to run
    it.

Version 0.95:
    Changed the Performance Data Field to only contain a label and the
    actual value in celsius. For example, "temperature=39".

Version 1.0:
    Several improvements made by Nikodimos, including:
     - Added support for multiple sensors
     - Enriched performance data output. For example, "CPU=34;40;55 MB=32;40;55"
     - Changed (and improved) the way temperatures are retrieved
     - Other fixes and touch-ups, such as the help text now shows long options

Version 1.1:
	Improvements for Icinga 2, made by Onkobu
	 - order of arguments is irrelevant now, --sensor first ignored following -c & -w
	 - current sensors' multiline output needs sophisticated treatment

Version 1.2:
   Improvements for Icinga 2, made by Onkobu
   - Added the ability to switch between old and new style with an -n option
   - Fixed multi-sensor with identical thresholds

Version 1.3:
	Fixes for multi sensor, made by Onkobu
	 - each sensor appears exactly once in message and in performance metrics
	Test infrastructure
	 - some Bash scripts and sample output mocking sensors' output
	 - outlined some terms (device, adapter, sensor) in README
	 - short explanation how to add tests in README



tests/run_test.sh
#!/bin/bash
export esc="\e"
export COLOR_RST="$esc[0m"
export COLOR_RED="$esc[31m"
export COLOR_GRN="$esc[32m"
export COLOR_YLW="$esc[33m"
export COLOR_LGREEN="$esc[1;32m" 

PATTERN='%-25s %-45s %s'
totalTests=0
testsFailed=0
testsSucceeded=0

for testCase in $(ls ./test_*.sh); do
	(( totalTests++ ))
	echo "*** Start testcase $testCase"
	$testCase
	errCode=$?
	
	if [ $errCode -ne 0 ]; then
		(( testsFailed++ ))
		status="${COLOR_RED}failed${COLOR_RST}"
	else
		(( testsSucceeded++ ))
		status="${COLOR_LGREEN}success${COLOR_RST}"
	fi

	printf -v msg "$PATTERN" $testCase "->" $status

	echo -e "$msg"
	echo -e "----------------------------------------\n"
done

echo Of $totalTests tests $testsFailed failed and $testsSucceeded succeeded


tests/test_relabel.sh
#!/bin/bash
. common.inc

export TESTDATA=$(sample_from_scriptname $0)

result=$(../check_temp.sh -w 65 -c 75 --sensor CPU,Emily)

matcher2=$(has_stats "$result" 'Emily=44;65;75')

exit $(( matcher1 + matcher2 ))



tests/test_new_filter.sh
#!/bin/bash
. common.inc

export TESTDATA_PREFIX=$(sample_from_scriptname $0)

echo $TESTDATA

# -n enables filtering by device through sensors command
#
result=$(../check_temp.sh -n -w 65 -c 75 --sensor k10temp-pci-00c3,CPU -w 60 -c 80 --sensor radeon-pci-0008,GPU)

matcher1=$(has_temperature "$result" '+45.2°C,' '+40.7°C')
matcher2=$(has_stats "$result" 'CPU=45;65;75 GPU=40;60;80')

exit $(( matcher1 + matcher2 ))



check_temp.sh
#!/bin/bash

################################################################################
#                                                                              #
#  Copyright (C) 2011 Jack-Benny Persson <jake@cyberinfo.se>                   #
#                                                                              #
#   This program is free software; you can redistribute it and/or modify       #
#   it under the terms of the GNU General Public License as published by       #
#   the Free Software Foundation; either version 2 of the License, or          #
#   (at your option) any later version.                                        #
#                                                                              #
#   This program is distributed in the hope that it will be useful,            #
#   but WITHOUT ANY WARRANTY; without even the implied warranty of             #
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              #
#   GNU General Public License for more details.                               #
#                                                                              #
#   You should have received a copy of the GNU General Public License          #
#   along with this program; if not, write to the Free Software                #
#   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA  #
#                                                                              #
################################################################################

###############################################################################
#                                                                             #	
# Nagios plugin to monitor CPU and M/B temperature with sensors.              #
# Written in Bash (and uses sed & awk).                                       #
# Latest version of check_temp can be found at the below URL:                 #
# https://github.com/jackbenny/check_temp                                     #
#                                                                             #
# If you are having problems getting it to work, check the instructions in    #
# the README first. It walks you though install lm-sensors and getting it to  #
# display sensor data.                                                        #
#                                                                             #
###############################################################################

VERSION="Version 1.3"
AUTHOR="(c) 2011 Jack-Benny Persson (jack-benny@cyberinfo.se), (c) 2020 Onkobu Tanaake (oss@onkobutanaake.de)"

# Sensor program
SENSORPROG=$(whereis -b -B /{bin,sbin,usr} /{bin,sbin,usr}/* -f sensors | awk '{print $2}')

# Ryan's note: utils.sh is installed with nagios-plugins in with the plugins
# Check if utils.sh exists. This lets you use check_domain in a testing environment
# or outside of Nagios.
if [ -e "$PROGPATH/utils.sh" ]; then
	. "$PROGPATH/utils.sh"
else
	STATE_OK=0
	STATE_WARNING=1
	STATE_CRITICAL=2
	STATE_UNKNOWN=3
#	STATE_DEPENDENT=4    (Commented because it's unused.)
fi

shopt -s extglob

#### Functions ####

# Print version information
print_version()
{
	echo "$0 - $VERSION"
}

#Print help information
print_help()
{
	print_version
	echo "$AUTHOR"
	echo "Monitor temperature with the use of sensors"
/bin/cat <<EOT

Options:
-h, --help
   Print detailed help screen
-V, --version
   Print version information
-v, --verbose
   Verbose output

-s, --sensor <WORD[,DISPLAY_NAME]>
   Set what to monitor, for example CPU or MB (or M/B). Check sensors for the
   correct word. Default is CPU. A different display name can be used in output,
   by adding it next to sensor with a comma.
   It can be used more than once, with different warning/critical thresholds optionally.
-w, --warning <INTEGER>
   Exit with WARNING status if above INTEGER degrees
-c, --critical <INTEGER>
   Exit with CRITICAL status if above INTEGER degrees
   Warning and critical thresholds must be provided before the corresponding --sensor option.
-n
   Use the new sed based filter in case classic filter yields no temperature.

Examples:
./check_temp.sh [-n] -w 65 -c 75 --sensor CPU
./check_temp.sh [-n] -w 65 -c 75 --sensor CPU --sensor temp1
./check_temp.sh [-n] -w 65 -c 75 --sensor CPU -w 75 -c 85 --sensor temp1,GPU
EOT
}


###### MAIN ########

# Warning threshold
thresh_warn=
# Critical threshold
thresh_crit=
# Hardware to monitor
default_sensor="CPU"
sensor_declared=false

STATE=$STATE_OK

# See if we have sensors program installed and can execute it
if [ $SENSORPROG = 'sensors' ] && [ "$(LC_ALL=C type -t $SENSORPROG)" = "function" ]; then
	echo "!!! TEST MODE !!!"
elif [[ ! -x "$SENSORPROG" ]]; then
	echo "It appears you don't have lm-sensors installed. You may find help in the readme for this script."
	exit $STATE_UNKNOWN
fi

function set_state {
	[[ "$STATE" -lt "$1" ]] && STATE=$1 
}

function process_sensor {
	sensor=$(echo $1 | cut -d, -f1)
	sensor_display=$(echo $1 | cut -d, -f2)
	# Check if a sensor were specified
	if [[ -z "$sensor" ]]; then
		# No sensor to monitor were specified
		echo "No sensor specified"
		print_help
		exit $STATE_UNKNOWN
	fi

	# Check if the thresholds have been set correctly
	if [[ -z "$thresh_warn" || -z "$thresh_crit" ]]; then
		# One or both thresholds were not specified
		echo "Threshold not set"
		print_help
		exit $STATE_UNKNOWN
	  elif [[ "$thresh_crit" -lt "$thresh_warn" ]]; then
		# The warning threshold must be lower than the critical threshold
		echo "Warning temperature should be lower than critical"
		print_help
		exit $STATE_UNKNOWN
	fi
	# Get the temperature
	# Grep the first float with a plus sign and keep only the integer
	if [ $CLASSIC_FILTER -eq 1 ]; then
		WHOLE_TEMP=$(${SENSORPROG} | grep "$sensor" | head -n1 | grep -o "+[0-9]\+\(\.[0-9]\+\)\?[^ \t,()]*" | head -n1)
	else
		WHOLE_TEMP=$(${SENSORPROG} -A "$sensor" | sed -n '2 p' | grep -o "+[0-9]\+\(\.[0-9]\+\)\?[^ \t,()]*" | head -n1)
	fi
	TEMPF=$(echo "$WHOLE_TEMP" | grep -o "[0-9]\+\(\.[0-9]\+\)\?")
	TEMP=$(echo "$TEMPF" | cut -d. -f1)

	# Verbose output
	if [[ "$verbosity" -ge 1 ]]; then
	   /bin/cat <<__EOT
	Debugging information:
	  Warning threshold: $thresh_warn 
	  Critical threshold: $thresh_crit
	  Verbosity level: $verbosity
	  Current $sensor temperature: $TEMP
__EOT
	echo "Temperature lines directly from sensors:"
	${SENSORPROG}
	fi
	
	# Get performance data for Nagios "Performance Data" field
	PERFDATA="$PERFDATA $sensor_display=$TEMP;$thresh_warn;$thresh_crit"

	# And finally check the temperature against our thresholds
	if [[ "$TEMP" != +([0-9]) ]]; then
		# Temperature not found for that sensor
		OUTPUT_TEXT="$OUTPUT_TEXT, No data found for sensor ($sensor)"
		set_state $STATE_UNKNOWN
	elif [[ "$TEMP" -gt "$thresh_crit" ]]; then
		# Temperature is above critical threshold
		OUTPUT_TEXT="$OUTPUT_TEXT, $sensor_display has temperature: $WHOLE_TEMP"
		set_state $STATE_CRITICAL
	elif [[ "$TEMP" -gt "$thresh_warn" ]]; then
		# Temperature is above warning threshold
		OUTPUT_TEXT="$OUTPUT_TEXT, $sensor_display has temperature: $WHOLE_TEMP"
		set_state $STATE_WARNING
	else
		# Temperature is ok
		OUTPUT_TEXT="$OUTPUT_TEXT, $sensor_display has temperature: $WHOLE_TEMP"
		set_state $STATE_OK
	fi
}

CLASSIC_FILTER=1

# Parse command line options
while [[ -n "$1" ]]; do 
   case "$1" in

       -h | --help)
           print_help
           exit $STATE_OK
           ;;

       -V | --version)
           print_version
           exit $STATE_OK
           ;;

       -v | --verbose)
           : $(( verbosity++ ))
           shift
           ;;

       -w | --warning)
           if [[ -z "$2" ]]; then
               # Threshold not provided
               echo "Option $1 requires an argument"
               print_help
               exit $STATE_UNKNOWN
            elif [[ "$2" = +([0-9]) ]]; then
               # Threshold is an integer 
               thresh=$2
            else
               # Threshold is not an integer
               echo "Threshold must be an integer"
               print_help
               exit $STATE_UNKNOWN
           fi
           thresh_warn=$thresh
	   shift 2
           ;;

       -c | --critical)
           if [[ -z "$2" ]]; then
               # Threshold not provided
               echo "Option '$1' requires an argument"
               print_help
               exit $STATE_UNKNOWN
            elif [[ "$2" = +([0-9]) ]]; then
               # Threshold is an integer 
               thresh=$2
            else
               # Threshold is not an integer
               echo "Threshold must be an integer"
               print_help
               exit $STATE_UNKNOWN
           fi
           thresh_crit=$thresh
	   shift 2
           ;;

       -s | --sensor)
	   if [[ -z "$2" ]]; then
		echo "Option $1 requires an argument"
		print_help
		exit $STATE_UNKNOWN
	   fi
	   sensor_declared=true
	   sensors_to_check="$2"
           shift 2
           ;;

	   -n | --new-filter)
		   CLASSIC_FILTER=0
		   shift 1
		   ;;
       *)
           echo "Invalid option '$1'"
           print_help
           exit $STATE_UNKNOWN
           ;;
   esac

   # argument order is irrelevant, Icinga2 gives no guarantees
   # as soon as there are enough output is generated
   if [ ! -z "$thresh_warn" ] && [ ! -z "$thresh_crit" ] && [ ! -z "$sensors_to_check" -o $# -eq 0 ]; then
	if [ "$sensor_declared" = false ]; then
		process_sensor "$default_sensor"
	else
		process_sensor "$sensors_to_check"
		unset sensors_to_check
	fi
	unset thresh_warn
	unset thresh_crit
   fi
done


case "$STATE" in
	"$STATE_OK") STATE_TEXT="OK" ;;
	"$STATE_WARNING") STATE_TEXT="WARNING" ;;
	"$STATE_CRITICAL") STATE_TEXT="CRITICAL" ;;
	"$STATE_UNKNOWN") STATE_TEXT="UNKNOWN" ;;
esac

OUTPUT_TEXT=$(echo $OUTPUT_TEXT | sed -e 's/, //')
echo "TEMPERATURE $STATE_TEXT - $OUTPUT_TEXT |$PERFDATA"
exit $STATE
